### PR TITLE
Add setting to refresh directory contents

### PR DESCRIPTION
**Pull request by bikemike**
Refresh a directory's contents when fs_readdir is called if it has been more than the number of seconds specified by --refresh_timeout since the directory was last indexed.

----

I haven't tested this with the cache option but thought I'd send along in case it was useful to anyone else. Sometimes it's nice to see new files in a directory without having to restart httpdirfs.

### CODE REVIEW

File: src/link.c [hunk 1]
```diff
@@ -733,24 +742,50 @@ LinkTable *LinkTable_disk_open(const char *dirn)
 
 LinkTable *path_to_Link_LinkTable_new(const char *path)
 {
-    Link *link = path_to_Link(path);
-    LinkTable *next_table = link->next_table;
+    struct Link *link = NULL, *tmp_link = NULL;
```
Comment (fangfufu) [inline src/link.c:745]: What are all these changes about?

File: src/fuse_local.c [hunk 2]
```diff
@@ -138,15 +138,12 @@ fs_readdir(const char *path, void *buf, fuse_fill_dir_t dir_add,
     (void) fi;
     LinkTable *linktbl;
 
-    if (!strcmp(path, "/")) {
```
Comment (fangfufu) [inline src/fuse_local.c:141]: What's the rationale behind the changes here?

**Review by fangfufu — COMMENTED** [review]
I would like some rationale behind the changes unrelated to refreshing the directory content - I commented on the relevant sections. If I am happy with the explanations, I will merge these.

**Comment by bikemike** [comment]
> I would like some rationale behind the changes unrelated to refreshing the directory content - I commented on the relevant sections. If I am happy with the explanations, I will merge these.

Those changes were made to treat the root directory as any other directory and thus allow its contents to be refreshed.

The changes marked move the special handling of the root directory and its associated variable(ROOT_LINK_TBL) from fs_readdir to path_to_Link_LinkTable_new and allows that variable to be updated when the directory should be refreshed.

**Comment by fangfufu** [comment]
Done, I have merged them in. :)

File: src/link.c [hunk 3]
```diff
@@ -733,24 +742,50 @@ LinkTable *LinkTable_disk_open(const char *dirn)
...
+    struct Link linkcpy = {0};
+    LinkTable *next_table = NULL;
+    if (!strcmp(path, "/")) {
+        next_table = ROOT_LINK_TBL;
+        linkcpy = *next_table->links[0];
+        tmp_link = &linkcpy;
+    } else {
+        link = path_to_Link(path);
+        tmp_link = link;
+        LinkTable *next_table = link->next_table;
```
Comment (fangfufu) [inline src/link.c:755]: I am pretty sure this line does absolutely nothing, because the shadowed variable goes out of scope immediately.

Comment (fangfufu) [reply src/link.c:755]: Yup the compiler says the variable is unused....

File: src/link.c [hunk 4]
```diff
@@ -501,11 +501,15 @@ static void LinkTable_invalid_reset(LinkTable *linktbl)
 
 void LinkTable_free(LinkTable *linktbl)
 {
-    for (int i = 0; i < linktbl->num; i++) {
-        FREE(linktbl->links[i]);
+    if (linktbl)
```
Comment (fangfufu) [inline src/link.c:504]: The pointer to a freed linktbl might not be ``NULL``...

File: src/link.c (hunk 1 above)
Comment (fangfufu) [inline src/link.c:745]: Comma operator should not be used. 
https://rules.sonarsource.com/c/RSPEC-878/

Comment (fangfufu) [reply src/link.c:745]: Also I have typedef-ed Link, you didn't need to use ``struct``

Comment (bikemike) [reply src/link.c:745]: You have already merged these changes. Are you going to update your repo or are you wanting me to create another PR?

File: src/link.c (hunk 4 above)
Comment (fangfufu) [reply src/link.c:504]: My point is that if you could potentially free a LinkTable that is already freed.

File: src/link.c (hunk 1 above)
Comment (fangfufu) [reply src/link.c:745]: I am just going to update the changes myself. I have just found myself some time to work on this project.

File: src/link.c [hunk 5]
```diff
@@ -733,24 +742,50 @@ LinkTable *LinkTable_disk_open(const char *dirn)
...
 LinkTable *path_to_Link_LinkTable_new(const char *path)
 {
-    Link *link = path_to_Link(path);
-    LinkTable *next_table = link->next_table;
+    struct Link *link = NULL, *tmp_link = NULL;
+    struct Link linkcpy = {0};
+    LinkTable *next_table = NULL;
+    if (!strcmp(path, "/")) {
+        next_table = ROOT_LINK_TBL;
+        linkcpy = *next_table->links[0];
```
Comment (fangfufu) [inline src/link.c:750]: I am having problem of understanding what ``*next_table->links[0];`` is.

File: src/link.c (hunk 4 above)
Comment (bikemike) [reply src/link.c:504]: Do my changes create this potential or could that happen regardless of the changes I made? It seems that this could happen with the original code too.

Comment (fangfufu) [reply src/link.c:504]: > Do my changes create this potential or could that happen regardless of the changes I made?

Your change increases this potential, because I didn't go around to the ``LinkTable``s to systematically free everything.

Comment (fangfufu) [reply src/link.c:504]: By the look of it, everything should work. Your ``LinkTable_free`` improves on my previous implementation. I am going to leave this for now. I think I should really write tests for this filesystem and run them on GitHub workflow.

**Comment by fangfufu** [comment]
Just for the record, this commit never handled the saving / loading aspects of the ``index_time``, this is causing https://github.com/fangfufu/httpdirfs/issues/140
